src/path.cr

```diff
@@ -351,49 +351,30 @@ struct Path
   # Path["foo.tar.gz"].extension # => ".gz"
   # ```
   def extension : String
+    return "" if @name.bytesize < 3
     bytes = @name.to_slice
+    separators = separators().map &.ord
 
-    return "" if bytes.empty?
-
-    slice_end = bytes.size
-    current = slice_end - 1
-
-    separators = self.separators.map &.ord
-
-    # ignore trailing separators
-    while separators.includes?(bytes[current]) && current > 0
-      current -= 1
-      slice_end -= 1
+    # Ignore trailing seperators
+    offset = bytes.size - 1
+    while bytes.unsafe_fetch(offset).in? separators
+      return "" if offset == 0
+      offset -= 1
     end
 
-    # if the pattern is `foo.`, it has no extension
-    return "" if bytes[current] == '.'.ord
-
-    # position the reader at the last `.` or SEPARATOR
-    # that is not the first char
-    while !separators.includes?(bytes[current]) &&
-          bytes[current] != '.'.ord &&
-          current > 0
-      current -= 1
-    end
-
-    # if we are the beginning of the string there is no extension
-    # `/foo` and `.foo` have no extension
-    return "" unless current > 0
-
-    # otherwise we are not at the beginning, and there is a previous char.
-    # if current is '/', then the pattern is prefix/foo and has no extension
-    return "" if separators.includes?(bytes[current])
+    # Get the first occurance of a seperator or a '.' past the trailing seperators (or return "")
+    dot_index = bytes.rindex(offset: offset) { |byte| byte === '.' || byte.in? separators } || return ""
 
-    # otherwise the current_char is '.'
-    # if previous is '/', then the pattern is `prefix/.foo`  and has no extension
-    return "" if separators.includes?(bytes[current - 1])
+    # Return "" if '.' is the first character (ex. ".dotfile"),
+    # or if the '.' character follows after a seperator (ex. "pathto/.dotfile")
+    # or if the character at the returned index is a seperator (ex. "no/extension")
+    # or if the filename ends with a '.'
+    return "" if dot_index == 0 ||
+                 dot_index == offset ||
+                 bytes.unsafe_fetch(dot_index - 1).in?(separators) ||
+                 bytes.unsafe_fetch(dot_index).in?(separators)
```

Comment:
Did you verify that combining all conditions into a disjunction is actually more performant than separate conditionals and return expressions?

Reply:
This was actually not done because of performance.
I just thought it looked nicer, but I could of course change it to four seperate return statements.

Reply:
We'd like to avoid changing from one style to another just for aesthetics. It's subjective anyways and doesn't really add anything except noise.
Performance optimization is a reason to change something. Improving readability of the code is also a reason. But in this case, I don't see a structural improvement. I'd even say it's a degradation because of the visual distance between conditions and respective comments.